llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp

```diff
@@ -3963,8 +3963,21 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
         Elements.push_back(DAG.getCTSelect(dl, ScalarVT, Tmp1, TVal, FVal, Node->getFlags()));
       }
       Tmp1 = DAG.getBuildVector(VT, dl, Elements);
+    } else if (VT.isFloatingPoint()) {
+      EVT IntegerVT = EVT::getIntegerVT(*DAG.getContext(), VT.getSizeInBits());
+      Tmp2 = DAG.getBitcast(IntegerVT, Tmp2);
+      Tmp3 = DAG.getBitcast(IntegerVT, Tmp3);
+      Tmp1 = DAG.getBitcast(VT, DAG.getCTSelect(dl, IntegerVT, Tmp1, Tmp2, Tmp3, Node->getFlags()));
     } else {
-      Tmp1 = DAG.getCTSelect(dl, Tmp1.getValueType(), Tmp1, Tmp2, Tmp3);
+      assert(VT.isInteger());
+      EVT HalfVT = VT.getHalfSizedIntegerVT(*DAG.getContext());
```

Comment:
Why do you need to split here? Seems it is dealing with 64 bits and getting split to half size. Aarch64 does support 64 bit operation. Is it possible to handle with them?

Reply:
This is not needed for AArch64 currently. The idea is that we can support CTSELECT on larger integer sizes than natively supported by splitting the CTSELECT into two until the right size is found. E.g. a CTSELECT on `i128` could be expanded into two `CTSELECT` of size `i64`

Reply:
Wait, don't bigger integers than i32 get handled in fp registers anyway? That is at least the case for ARM - maybe not for AArch64?

Reply:
Make sense for i128 but should we add more checks around it. I do see someone intend to use Expand action on types MVT::i64 for 32 bit target. It will work but the code could be common for both 32 and 64 bit variant unless some used sub-target checks.

Also, I believe all target does not support `i128` so should it left to target lowering for handling the specific case?

Reply:
The Expand action is set in target specific lowering construct. That does addresses second part of my comment where there are operands and register to handle 128 bits. 

This change looks good to me. The action should be set according to bitwidth and if subtarget is 64 bit or not.

Reply:
> Make sense for i128 but should we add checks for the bitwidth? I do see someone intend to use Expand action on types MVT::i64 for 32 bit target. It will work but the code could be common for both 32 and 64 bit variant unless some used sub-target checks.

Sorry, I don't understand this specific question...

Reply:
My concern is related to targets that have common lowering pipeline for both 32 and 64 bits variant. But I guess it is responsibility of target to register for action based on whichever types they want to and they will add checks to handle it.

Reply:
If a target allows CTSELECTing 32 and 64 bits natively, you would do `setOperationAction(CTSELECT, MVT::i32, Custom)` and `setOperationAction(CTSELECT, MVT::i64, Custom)`, no need for expansion :)